src/java.base/share/classes/java/io/Console.java

```diff
@@ -145,6 +146,66 @@ public Reader reader() {
         throw newUnsupportedOperationException();
     }
 
+    /**
+     * Writes a string representation of the specified object to this console's
+     * output stream, terminates the line and then flushes the console.
+     *
+     * <p> The string representation of the specified object is obtained as if
+     * by calling {@link String#valueOf(Object)}.
+     *
+     * @param  obj
+     *         An object whose string representation is to be written.
+     *
+     * @return  This console
+     *
+     * @since 23
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
+    public Console println(Object obj) {
+        throw newUnsupportedOperationException();
+    }
+
+    /**
+     * Writes a string representation of the specified object to this console's
+     * output stream and then flushes the console.
+     *
+     * <p> The string representation of the specified object is obtained as if
+     * by calling {@link String#valueOf(Object)}.
+     *
+     * @param  obj
+     *         An object whose string representation is to be written.
+     *
+     * @return  This console
+     *
+     * @since 23
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
+    public Console print(Object obj) {
+        throw newUnsupportedOperationException();
+    }
+
+    /**
+     * Writes a prompt as if by calling {@code print}, then reads a single line
+     * of text from this console.
+     *
+     * @param  prompt
+     *         A prompt string.
```

Comment:
Hello Pavel, should this specify whether `prompt`  can be null?

Reply:
If we specify that, it would be very much unlike all other `Console` methods that are covered by this:

    * Unless otherwise specified, passing a {@code null} argument to any method
    * in this class will cause a {@link NullPointerException} to be thrown.

Reply:
I haven't given it a try, but a brief look at the code suggests that if the console implementation backed by JLine gets used, then a `null` prompt may not generate a `NullPointerException` (since JLine appears to allow `null`) whereas if the internal JDK console implementation gets used then a `NullPointerException` will get thrown. If the expectation is to disallow a `null` prompt, then perhaps `Objects.requireNonNull(prompt)` before we hand off to the underlying console implementations might be required.

Reply:
Good catch! `jdk.internal.org.jline.JdkConsoleProviderImpl.JdkConsoleImpl.readln(null)` does not throw `NullPointerException` (NPE), which it must. I'll fix the bug and add a test.

Speaking of NPE. The newly added `Console.print` and `Console.println` methods do not throw NPE if passed null. While that's how it should be, they don't specify that, so the class-level NPE clause applies:

    * Unless otherwise specified, passing a {@code null} argument to any method
    * in this class will cause a {@link NullPointerException} to be thrown.
    ...
    public sealed class Console implements Flushable permits ProxyingConsole {

I'll fix that.

We should also specify NPE (or absence thereof) on methods of `IO`. Otherwise, this package clause applies:

    * Unless otherwise noted, passing a {@code null} argument to a constructor or
    * method in any class or interface in this package will cause a
    * {@code NullPointerException} to be thrown.
    ...
    package java.io;

@stuart-marks speaking of exceptions. Do we need to add explicit `@throws` tag to `IO` methods?

    @throws {@code IOError} if {@code System.console()} returns {@code null}

While it would make sense, it would also look superfluous: we already specify that error condition in the class comment and then in every method's main description. Do we need to add one more note on `IOError`?

Reply:
Most of these are defined in terms of `String.valueOf(Object)` which if passed a null reference will return the String object containing `null`. So no, I don't think NPE should be thrown. It might be worth mentioning this explicitly somewhere though.

Reply:
> Most of these are defined in terms of `String.valueOf(Object)` which if passed a null reference will return the String object containing `null`. So no, I don't think NPE should be thrown. It might be worth mentioning this explicitly somewhere though.

Right, in our current model, `Console.readln(prompt)` can be reduced to `Console.print(prompt)` followed by actual reading. `Console.print(prompt)` can in turn be reduced to printing the result of `String.valueOf((Object) prompt)`, which is a string "null".

The above reduction means that `Console.readln(null)` will not throw NPE. Now, what I'm asking if that's what we really want.

With `Console.println(null)` and `Console.print(null)` not throwing NPE makes sense. However, with `Console.readln(null)` it might not. After all, we ask for a **`String`** prompt, not an `Object` prompt. All other methods of `Console` that accept `String`, throw NPE if that string is null.

One more consideration. Whatever the behaviour we choose for `Console.readln(null)`, we should choose the same behaviour for `IO.readln(null)`. Otherwise, it would be a constant source of unpleasantly surprising inconsistency.

Thoughts?

Reply:
A fair question, whether to treat `readln` differently because it takes a String argument instead of an Object.

I guess I'd like to see a scheme like the following:

    String s = cons.readln(prompt);

is effectively the same as

    cons.print(prompt);
    cons.flush();
    String s = cons.readLine(); // note no-arg readLine() method

In other words, define `readln` to behave "as if" the other methods were called. This would end up with a null prompt argument printing "null" and not throwing NPE.

This is internally consistent, at least. Beginning programmers will eventually learn that using null will lead to NPE. It might make things easier if NPEs were kept out of this little corner of the library. That's just my guess though.

Also, recall that `String.valueOf((Object) null)` and `String.valueOf((String) null)` both result in "null". But if one were to write `String.valueOf(null)` that throws NPE, because that selects the `valueOf(char[])` overload. That's kind of the outlier. Note further that the built-in `+` string concatenation operator will also convert null references to "null".

Personally I've always hated that string conversions of null in Java often result in "null" but it's baked into the system fairly deeply, so I think we should stick with it.

Reply:
Okay, I'll push a commit soon.